File: solutions/33 - Search in Rotated Sorted Array/Binary Search.java
import java.util.*;
class Solution {
    public int search(int[] nums, int target) {
        // Time Complexity: O(LogN) to binary search.
        // Space Complexity: O(1) for a few variables.
        int l = 0;
        int r = nums.length - 1;
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (nums[m] == target) {
                return m;
            }
            if (nums[m] >= nums[l]) {
                if (target >= nums[l] && target < nums[m]) {
                    r = m - 1;
                } else {
                    l = m + 1;
                }
            } else {
                if (target <= nums[r] && target > nums[m]) {
                    l = m + 1;
                } else {
                    r = m - 1;
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
    
        System.out.println("Enter nums: ");
        int[] nums = Arrays.stream(sc.nextLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();
    
        System.out.println("Enter target: ");
        int target = sc.nextInt();
    
        System.out.println("Index: ");
        System.out.println(new Solution().search(nums, target));
    
        sc.close();
    }
}

File: solutions/33 - Search in Rotated Sorted Array/Binary Search.cpp
#include <iostream>
#include <vector>
#include <sstream>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        // Time Complexity: O(LogN) to binary search.
        // Space Complexity: O(1) for a few variables.
        int low = 0, high = nums.size() - 1;
        while (low <= high)
        {
            int mid = low + (high - low) / 2;
            if (nums[mid] == target)
            {
                return mid;
            }
            // left half sorted
            if (nums[low] <= nums[mid])
            {
                if (target < nums[mid] && target >= nums[low])
                {
                    high = mid - 1;
                }
                else
                {
                    low = mid + 1;
                }
            }
            // right half sorted
            else
            {
                if (target > nums[mid] && target <= nums[high])
                {
                    low = mid + 1;
                }
                else
                {
                    high = mid - 1;
                }
            }
        }
        
        return -1;
    }
};

int main() {
    Solution sol;
    vector<int> nums;
    string line;
    int target;

    cout << "Enter nums: " << endl;
    getline(cin, line);

    stringstream ss(line);
    int x;
    while (ss >> x) {
        nums.push_back(x);
    }

    cout << "Enter target: " << endl;
    cin >> target;

    int result = sol.search(nums, target);

    cout << "Index:" << endl;
    cout << result << endl;

    return 0;
}